File: src/core/schema.ts
export interface ContextLaneSource {
  id: string
  type: 'file' | 'folder' | 'url' | 'github' | 'text'
  input: string
  title?: string
  path?: string
  url?: string
  hash?: string
  loadedAt: string
  metadata?: Record<string, unknown>
}

export interface ContextLaneChunk {
  id: string
  sourceId: string
  text: string
  index: number
  startLine?: number
  endLine?: number
  citation: string
  metadata?: Record<string, unknown>
}

export interface ContextLaneFact {
  id: string
  text: string
  confidence: number
  tags: string[]
  citationIds: string[]
}

export interface ContextLaneDecision {
  id: string
  text: string
  status: 'active' | 'historical' | 'unknown'
  citationIds: string[]
}

export interface ContextLaneAction {
  id: string
  text: string
  priority: 'low' | 'medium' | 'high'
  citationIds: string[]
}

export interface ContextLaneEntity {
  id: string
  name: string
  type: 'person' | 'project' | 'tool' | 'company' | 'url' | 'concept' | 'other'
  mentions: number
  citationIds: string[]
}

export interface ContextLaneCitation {
  id: string
  sourceId: string
  label: string
  path?: string
  url?: string
  startLine?: number
  endLine?: number
  quote?: string
}

export interface ContextLaneExtraction {
  runId: string
  summary: string
  facts: ContextLaneFact[]
  decisions: ContextLaneDecision[]
  actions: ContextLaneAction[]
  entities: ContextLaneEntity[]
  tags: string[]
  citations: ContextLaneCitation[]
  createdAt: string
}

export interface ContextLaneMemoryRecord {
  text: string
  tags: string[]
  source: string
  citation: string
  metadata: {
    runId: string
    sourceId: string
    citationIds: string[]
    contextlane: true
  }
  importance: number
}

export interface ContextLaneRun {
  id: string
  sources: ContextLaneSource[]
  chunks: ContextLaneChunk[]
  extraction: ContextLaneExtraction
  memoryRecords: ContextLaneMemoryRecord[]
  createdAt: string
}

export interface ContextLaneRunMeta {
  id: string
  sourceCount: number
  chunkCount: number
  factCount: number
  entityCount: number
  createdAt: string
}

File: python/contextlane/core/schema.py
from dataclasses import dataclass, field
from typing import Optional


@dataclass
class ContextLaneSource:
    id: str
    type: str  # file | folder | url | github | text
    input: str
    title: Optional[str] = None
    path: Optional[str] = None
    url: Optional[str] = None
    hash: Optional[str] = None
    loaded_at: str = ""
    metadata: Optional[dict] = None


@dataclass
class ContextLaneChunk:
    id: str
    source_id: str
    text: str
    index: int
    start_line: Optional[int] = None
    end_line: Optional[int] = None
    citation: str = ""
    metadata: Optional[dict] = None


@dataclass
class ContextLaneFact:
    id: str
    text: str
    confidence: float = 0.7
    tags: list = field(default_factory=list)
    citation_ids: list = field(default_factory=list)


@dataclass
class ContextLaneDecision:
    id: str
    text: str
    status: str = "active"  # active | historical | unknown
    citation_ids: list = field(default_factory=list)


@dataclass
class ContextLaneAction:
    id: str
    text: str
    priority: str = "medium"  # low | medium | high
    citation_ids: list = field(default_factory=list)


@dataclass
class ContextLaneEntity:
    id: str
    name: str
    type: str = "other"  # person | project | tool | company | url | concept | other
    mentions: int = 1
    citation_ids: list = field(default_factory=list)


@dataclass
class ContextLaneCitation:
    id: str
    source_id: str
    label: str
    path: Optional[str] = None
    url: Optional[str] = None
    start_line: Optional[int] = None
    end_line: Optional[int] = None
    quote: Optional[str] = None


@dataclass
class ContextLaneExtraction:
    run_id: str
    summary: str = ""
    facts: list = field(default_factory=list)
    decisions: list = field(default_factory=list)
    actions: list = field(default_factory=list)
    entities: list = field(default_factory=list)
    tags: list = field(default_factory=list)
    citations: list = field(default_factory=list)
    created_at: str = ""


@dataclass
class ContextLaneMemoryRecord:
    text: str
    tags: list = field(default_factory=list)
    source: str = ""
    citation: str = ""
    metadata: Optional[dict] = None
    importance: int = 5


@dataclass
class ContextLaneRun:
    id: str
    sources: list = field(default_factory=list)
    chunks: list = field(default_factory=list)
    extraction: Optional[ContextLaneExtraction] = None
    memory_records: list = field(default_factory=list)
    created_at: str = ""


@dataclass
class ContextLaneRunMeta:
    id: str
    source_count: int = 0
    chunk_count: int = 0
    fact_count: int = 0
    entity_count: int = 0
    created_at: str = ""
